File: algorithms/sorting-algorithms/mergesort.py
#Mergesort function which alsongside Merge function, performs
#a mergesort on the gven list
def MergeSort(list):
	if len(list) < 2:
		return list
	else:
		mid = len(list) / 2
		left = MergeSort(list[:mid])
		right = MergeSort(list[mid:])
		return Merge(left, right)
		
def Merge(left, right):
	merged = []
	left_index = 0
	right_index = 0
	while left_index < len(left) and right_index < len(right):
		if left[left_index] <= right[right_index]:
			merged.append(left[left_index])
			left_index += 1
		else:
			merged.append(right[right_index])
			right_index += 1
	if left:
		merged.extend(left[left_index:])
	if right:
		merged.extend(right[right_index:])
	return merged	

File: algorithms/sorting-algorithms/mergesort.java
package merge;
import java.util.*;

public class mergesort {
	
	public static void main(String[] args) {
		Integer[] a = {2, 6, 3, 5, 1};
		mergeSort(a);
		System.out.println(Arrays.toString(a));
	}
	
	public static void mergeSort(Comparable[] array) {
		Comparable[] temp = new Comparable[array.length];
		mergeSort(array, temp,  0,  array.length - 1);
	}
	
	private static void mergeSort(Comparable[] array, Comparable[] temp, int left, int right) {
		if (left < right) {
			int mid = (left + right) / 2;
			mergeSort(array, temp, left, mid);
			mergeSort(array, temp, mid+1, right);
			merge(array, temp, left, mid+1, right);
		}
	}
	
	private static void merge(Comparable[] array, Comparable[] temp, int left, int right, int rightEnd) {
		int leftEnd = right - 1;
		int k = left;
		int num = rightEnd - left + 1;
		
		while (left <= leftEnd && right <= rightEnd) {
			if (array[left].compareTo(array[right]) <= 0) {
				temp[k++] = array[left++];
			} else {
				temp[k++] = array[right++];
			}
		}
		
		while (left <= leftEnd) {
			temp[k++] = array[left++];
		}
		
		while (right <= rightEnd) {
			temp[k++] = array[right++];
		}
		
		for (int i = 0; i < num; i++, rightEnd--) {
			array[rightEnd] = temp[rightEnd];
		}
	}

}
